File: exercises/cpp/lab08/shm_writer.cpp
// Lab 08 - Shared Memory IPC (Writer)
#include <iostream>
#include <sys/ipc.h>
#include <sys/shm.h>
#include <cstring>
#include <cstdio>
using namespace std;

int main() {
    // Generate a unique key (creates "shmfile" if it doesn't exist)
    FILE* f = fopen("shmfile", "w"); fclose(f);
    key_t key = ftok("shmfile", 65);

    // Create or access shared memory segment
    int shmid = shmget(key, 1024, 0666 | IPC_CREAT);
    if (shmid == -1) { perror("shmget"); return 1; }

    // Attach to shared memory
    char* str = (char*)shmat(shmid, nullptr, 0);
    if (str == (char*)-1) { perror("shmat"); return 1; }

    cout << "Write data to shared memory: ";
    fgets(str, 100, stdin);

    // Remove trailing newline
    int len = strlen(str);
    if (len > 0 && str[len-1] == '\n') str[len-1] = '\0';

    printf("Data written to shared memory: \"%s\"\n", str);
    printf("Run shm_reader now to read the data.\n");

    shmdt(str);
    return 0;
}

File: exercises/python/lab08/shm_writer.py
"""Lab 08 - Shared Memory IPC (Writer) using mmap + file"""
import mmap
import os
import struct

SHM_FILE = "/tmp/os_lab08_shm"
SHM_SIZE = 1024


def main():
    data = input("Write data to shared memory: ").encode()
    # Pad to SHM_SIZE
    data = data[:SHM_SIZE-4].ljust(SHM_SIZE-4, b'\x00')
    length = len(data.rstrip(b'\x00'))

    with open(SHM_FILE, "wb") as f:
        f.write(struct.pack("I", length))  # 4 bytes for length
        f.write(data)

    print(f'Data written to shared memory: "{data[:length].decode()}"')
    print("Run shm_reader.py to read the data.")


if __name__ == "__main__":
    main()
